batch2/batch/driver/instance_pool.py

```diff
@@ -87,6 +87,9 @@ def adjust_for_remove_instance(self, instance):
             self.healthy_instances_by_free_cores.remove(instance)
 
     async def remove_instance(self, instance):
+        if instance._state == 'deleted':
+            return
```

Comment:
This isn't right.  My naming is perhaps a bit subtle: deleted means we (or someone) asked Google to delete the instance.  We don't remove the instance while the instance delete is in progress (it accounts against max instances).  Remove means the the Google instance delete succeeded (or we just noticed the instance is gone) and we should actually remove the instance from our list of instances (it no longer counts against max instances).

Reply:
ok. The issue is this function was probably called twice for the same instance.

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/batch/driver/instance_pool.py", line 392, in check_on_instance
    await session.get(f'http://{instance.ip_address}:5000/healthcheck')
  File "/usr/local/lib/python3.6/dist-packages/aiohttp/client.py", line 583, in _request
    break
  File "/usr/local/lib/python3.6/dist-packages/aiohttp/helpers.py", line 585, in __exit__
    raise asyncio.TimeoutError from None
concurrent.futures._base.TimeoutError
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/batch/driver/instance_pool.py", line 392, in check_on_instance
    await session.get(f'http://{instance.ip_address}:5000/healthcheck')
  File "/usr/local/lib/python3.6/dist-packages/aiohttp/client.py", line 583, in _request
    break
  File "/usr/local/lib/python3.6/dist-packages/aiohttp/helpers.py", line 585, in __exit__
    raise asyncio.TimeoutError from None
concurrent.futures._base.TimeoutError
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/batch/driver/instance_pool.py", line 392, in check_on_instance
    await session.get(f'http://{instance.ip_address}:5000/healthcheck')
  File "/usr/local/lib/python3.6/dist-packages/aiohttp/client.py", line 583, in _request
    break
  File "/usr/local/lib/python3.6/dist-packages/aiohttp/helpers.py", line 585, in __exit__
    raise asyncio.TimeoutError from None
concurrent.futures._base.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/batch/driver/instance_pool.py", line 438, in instance_monitoring_loop
    await self.check_on_instance(instance)
  File "/usr/local/lib/python3.6/dist-packages/batch/driver/instance_pool.py", line 397, in check_on_instance
    await instance.incr_failed_request_count()
  File "/usr/local/lib/python3.6/dist-packages/batch/driver/instance.py", line 145, in incr_failed_request_count
    self.instance_pool.adjust_for_remove_instance(self)
  File "/usr/local/lib/python3.6/dist-packages/batch/driver/instance_pool.py", line 83, in adjust_for_remove_instance
    self.instances_by_last_updated.remove(instance)
  File "/usr/local/lib/python3.6/dist-packages/sortedcontainers/sortedset.py", line 476, in remove
    self._set.remove(value)
KeyError: <batch.driver.instance.Instance object at 0x7f595c3a0b38>
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/batch/driver/instance_pool.py", line 392, in check_on_instance
    await session.get(f'http://{instance.ip_address}:5000/healthcheck')
  File "/usr/local/lib/python3.6/dist-packages/aiohttp/client.py", line 583, in _request
    break
  File "/usr/local/lib/python3.6/dist-packages/aiohttp/helpers.py", line 585, in __exit__
    raise asyncio.TimeoutError from None
concurrent.futures._base.TimeoutError
```